src/ssd/ssd_shadow.c

```diff
@@ -0,0 +1,279 @@
+// SPDX-License-Identifier: GPL-2.0-only
+
+#include <assert.h>
+#include "common/scene-helpers.h"
+#include "labwc.h"
+#include "buffer.h"
+#include "ssd-internal.h"
+#include "theme.h"
+#include "view.h"
+#include <cairo.h>
+
+#define FOR_EACH_STATE(ssd, tmp) FOR_EACH(tmp, \
+	&(ssd)->shadow.active, \
+	&(ssd)->shadow.inactive)
+
+/*
+ * Implements point_accepts_input for a buffer which never accepts input
+ * because drop-shadows should never catch clicks!
+ */
+static bool
+never_accepts_input(struct wlr_scene_buffer *buffer, double *sx, double *sy)
+{
+	return false;
+}
+
+/*
+ * Setup transform and scale for shadow corner buffers. Cropping is applied if
+ * the window is not tall enough and the top and bottom corners would overlap,
+ * the amount to crop is controlled by overlap_px.  Cropping is applied after
+ * rotation so switch_axes should be true for the bottom-left and top-right
+ * corners to crop horizontally instead of vertically.
+ */
+static void
+corner_scale_crop(struct wlr_scene_buffer *buffer, int overlap_px,
+		int corner_size, bool switch_axes)
+{
+	int width = corner_size;
+	int height = corner_size - overlap_px;
+	/* Crop is applied before rotation so gets the axis flip */
+	struct wlr_fbox src_box = {
+		.x = switch_axes ? overlap_px : 0,
+		.y = switch_axes ? 0 : overlap_px,
+		.width = switch_axes ? height : width,
+		.height = switch_axes ? width : height,
+	};
+	wlr_scene_buffer_set_source_box(buffer, &src_box);
+	/* But scaling is applied after rotation so no axis flip */
+	wlr_scene_buffer_set_dest_size(buffer, width, height);
+}
```

Comment:
It seems that the overlap concerns also apply to width, if somebody makes the radius large and a window is narrow. Shouldn't the crop consider widthwise overlap as well?

Reply:
Theoretically this could happen but I haven't seen any cases in testing nor has anyone testing reported any, so I didn't think it was worth the extra complication to handle it.  Seems like the vertical case is only actually seen in practice because of window shading.

Reply:
The halves aren't equal, one is an issue which was quickly noticed by actual users, the other is still theoretical and to me seems very unlikely.  The time I can spend on this is not unlimited and there's still plenty more to do so I'd appreciate if we could prioritise, and we can make further improvements once we find out the wild ways users break this in practice.

Reply:
The logic is the same, you should just add another overlap parameter that applies to the width and replace the zeros in this function with the other parameter. I understand that your time is not infinite, but I'm also opposed to releasing a feature that is known to be defective for limiting cases *in the same manner as you're already fixing*.

Reply:
I'll have a look at fixing this.  Please can you let me know when you've finished reviewing so I can address any remaining issues in one go in order to avoid any more re-rewriting and multiple rounds of testing/checking.